lokimq/bt_serialize.h

```diff
@@ -909,5 +922,192 @@ class bt_dict_consumer : private bt_list_consumer {
     bt_dict_consumer consume_dict_consumer() { return consume_dict_data(); }
 };
 
+/// Currently this is not idiot-proof.  That is to say, the compiler won't catch everything and
+/// you technically can create invalid BEncoded strings.  Basically you can add list elements
+/// to a dict and dict elements to a list, and obviously that can end badly.  This may be fixed
```

Comment:
I find this interface a bit clunky because of this: the caller has to worry about not calling the wrong method to screw things up, which means the caller has to know the technical details of bt_dict.  It would be cleaner to split them up into bt_buffer_list_producer/bt_buffer_dict_producer to mirror bt_dict_consumer/bt_list_consumer (perhaps with a common private base class for the implementation functions), where things like `append_integer(T)` are only in the list version and `append_integer(string_view, T)` are only in the dict version.

Otherwise you basically have two distinct functionalities in one single class: there are one set of methods that you may only call if using it as a list, and another set that you may only call if using it as a dict, and this feels wrong.